File: app/src/main/java/com/omarea/vboot/BatteryService.java
package com.omarea.vboot;

import android.app.ActivityManager;
import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.SharedPreferences;
import android.os.IBinder;

import java.util.List;

public class BatteryService extends Service {
    public BatteryService() {
    }


    @Override
    public IBinder onBind(Intent intent) {
        // TODO: Return the communication channel to the service.
        throw new UnsupportedOperationException("Not yet implemented");
    }

    reciver_batterychanged batteryChangedReciver;

    @Override
    public void onCreate() {
        if (batteryChangedReciver == null) {
            //监听电池改变
            batteryChangedReciver = new reciver_batterychanged();
            //启动完成
            IntentFilter ACTION_BOOT_COMPLETED = new IntentFilter(Intent.ACTION_BOOT_COMPLETED);
            registerReceiver(batteryChangedReciver, ACTION_BOOT_COMPLETED);
            //电源连接
            IntentFilter ACTION_POWER_CONNECTED = new IntentFilter(Intent.ACTION_POWER_CONNECTED);
            registerReceiver(batteryChangedReciver, ACTION_POWER_CONNECTED);
            //电源断开
            IntentFilter ACTION_POWER_DISCONNECTED = new IntentFilter(Intent.ACTION_POWER_DISCONNECTED);
            registerReceiver(batteryChangedReciver, ACTION_POWER_DISCONNECTED);
            //电量变化
            IntentFilter ACTION_BATTERY_CHANGED = new IntentFilter(Intent.ACTION_BATTERY_CHANGED);
            registerReceiver(batteryChangedReciver, ACTION_BATTERY_CHANGED);
            //电量不足
            IntentFilter ACTION_BATTERY_LOW = new IntentFilter(Intent.ACTION_BATTERY_LOW);
            registerReceiver(batteryChangedReciver, ACTION_BATTERY_LOW);
        }
    }

    @Override
    public void onDestroy() {
        if (batteryChangedReciver != null) {
            unregisterReceiver(batteryChangedReciver);
            batteryChangedReciver = null;
        }
    }

    //服务是否在运行
    public static Boolean serviceIsRunning(Context context) {
        ActivityManager m = (ActivityManager) (context.getSystemService(Context.ACTIVITY_SERVICE));
        List<ActivityManager.RunningServiceInfo> serviceInfos = m.getRunningServices(5000);
        for (ActivityManager.RunningServiceInfo serviceInfo : serviceInfos) {
            if (serviceInfo.service.getPackageName().equals("com.omarea.vboot")) {
                if (serviceInfo.service.getClassName().equals("com.omarea.vboot.BatteryService")) {
                    return true;
                }
            }
        }
        return false;
    }
}

File: app/src/main/java/com/omarea/vtools/services/BatteryService.kt
package com.omarea.vtools.services

import android.app.ActivityManager
import android.app.Service
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.IBinder
import com.omarea.charger_booster.BatteryReceiver
import com.omarea.data_collection.EventBus
import com.omarea.data_collection.publisher.BatteryState

public class BatteryService : Service() {
    internal var batteryChangedReciver: BatteryState? = null

    override fun onBind(intent: Intent): IBinder? {
        return null
    }

    override fun onCreate() {
        if (batteryChangedReciver == null) {
            //监听电池改变
            batteryChangedReciver = BatteryState()
            //启动完成
            registerReceiver(batteryChangedReciver, IntentFilter(Intent.ACTION_BOOT_COMPLETED))
            //电源连接
            registerReceiver(batteryChangedReciver, IntentFilter(Intent.ACTION_POWER_CONNECTED))
            //电源断开
            registerReceiver(batteryChangedReciver, IntentFilter(Intent.ACTION_POWER_DISCONNECTED))
            //电量变化
            registerReceiver(batteryChangedReciver, IntentFilter(Intent.ACTION_BATTERY_CHANGED))
            //电量不足
            registerReceiver(batteryChangedReciver, IntentFilter(Intent.ACTION_BATTERY_LOW))

            // 充电控制模块
            EventBus.subscibe(BatteryReceiver(applicationContext))
        }
    }

    override fun onDestroy() {
        try {
            if (batteryChangedReciver != null) {
                unregisterReceiver(batteryChangedReciver)
                batteryChangedReciver = null
            }
        } catch (igoned: Exception) {

        }
    }

    companion object {
        //服务是否在运行
        fun serviceIsRunning(context: Context): Boolean {
            val m = context.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager?
            if (m != null) {
                val className = BatteryService::class.java.simpleName
                val serviceInfos = m.getRunningServices(5000)
                for (serviceInfo in serviceInfos) {
                    if (serviceInfo.service.packageName == context.packageName) {
                        if (serviceInfo.service.className.endsWith(className)) {
                            return true
                        }
                    }
                }
            }
            return false
        }

        //启动电池服务
        fun startBatteryService(context: Context): Boolean {
            try {
                val intent = Intent(context, BatteryService::class.java)
                context.startService(intent)
                return true
            } catch (ex: Exception) {
                return serviceIsRunning(context)
            }
        }
    }
}
